src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -0,0 +1,345 @@
+        type=int,
+        help="Workspace addressed by the permanent ID",
+    )
+
+    # Optional argument
+    optional.add_argument(
+        "--root_dir",
+        type=str,
+        default=loader_common_names.ROOT_DIR,
+        help="Root directory.",
+    )
+    optional.add_argument(
+        "--kb_base_url",
+        type=str,
+        default=loader_common_names.KB_BASE_URL_DEFAULT,
+        help="KBase base URL, defaulting to prod",
+    )
+    optional.add_argument(
+        "--workers",
+        type=int,
+        default=5,
+        help="Number of workers for multiprocessing",
+    )
+    optional.add_argument(
+        "--token_filepath",
+        type=str,
+        help="A file path that stores KBase token",
+    )
+    optional.add_argument(
+        "--keep_job_dir",
+        action="store_true",
+        help="Keep SDK job directory after download task is completed",
+    )
+
+    args = parser.parse_args()
+    if args.workspace_id <= 0:
+        parser.error(f"workspace_id needs to be > 0")
+    if args.workers < 1 or args.workers > cpu_count():
+        parser.error(f"minimum worker is 1 and maximum worker is {cpu_count()}")
+
+    (workspace_id, root_dir, kb_base_url, workers, token_filepath, keep_job_dir) = (
+        args.workspace_id,
+        args.root_dir,
+        args.kb_base_url,
+        args.workers,
+        args.token_filepath,
+        args.keep_job_dir,
+    )
+
+    uid = os.getuid()
+    username = os.getlogin()
+
+    job_dir = _make_job_dir(root_dir, loader_common_names.SDK_JOB_DIR, username)
+    output_dir = _make_output_dir(
+        root_dir, loader_common_names.SOURCE_DATA_DIR, SOURCE, workspace_id
+    )
+
+    proc = None
+    conf = None
+
+    try:
+        # start podman service
+        proc = loader_helper.start_podman_service(uid)
+    except Exception as e:
+        raise Exception("Podman service failed to start") from e
+    else:
+        # set up conf and start callback server
+        conf = Conf(job_dir, output_dir, workers, kb_base_url, token_filepath)
+        objs = list_objects(workspace_id, conf, FILTER_OBJECTS_NAME_BY)
+        output_dir_upas = set(os.listdir(output_dir))
+        download = []
+
+        for obj_info in objs:
+            upa = "{6}_{0}_{4}".format(*obj_info)
+            if upa in output_dir_upas and loader_helper.is_upa_info_complete(
+                output_dir, upa
+            ):
+                continue
+            # remove legacy upa_dir to avoid FileExistsError in hard link
+            dirpath = os.path.join(output_dir, upa)
+            if os.path.exists(dirpath) and os.path.isdir(dirpath):
+                shutil.rmtree(dirpath)
+            conf.queue.put([upa, obj_info])
+            download.append(upa)
+
+        for i in range(workers + 1):
+            conf.queue.put(None)
+
+        conf.pools.close()
+        conf.pools.join()
+
+        print(f"These upas {download} are downloaded from workspace into {output_dir}")
```

Comment:
This could be 300K upas. Do you really want to print that out? If you want to give that info to the user printing it to file would be better

Reply:
Updated!

Reply:
:+1: